File: CodeBattleNet/Command.cs
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2018 - 2020 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */
using System.Collections.Generic;

namespace TetrisClient
{
	public class Command
	{
		public static readonly Command LEFT = new Command("LEFT");
		public static readonly Command RIGHT = new Command("RIGHT");
		public static readonly Command DOWN = new Command("DOWN");
		public static readonly Command ROTATE_CLOCKWISE_90 = new Command("ACT");
		public static readonly Command ROTATE_CLOCKWISE_180 = new Command("ACT(2)");
		public static readonly Command ROTATE_CLOCKWISE_270 = new Command("ACT(3)");
		public static readonly Command SUICIDE = new Command("ACT(0,0)");

		private readonly List<string> _commandChain = new List<string>();

		private Command(string command)
		{
			_commandChain.Add(command);
		}

		private Command()
		{
		}

		public Command Then(Command command)
		{
			var chained = new Command();
			chained._commandChain.AddRange(_commandChain);
			chained._commandChain.AddRange(command._commandChain);
			return chained;
		}

		public override string ToString()
		{
			return string.Join(",", _commandChain);
		}
	}
}

File: CodeBattleJava/src/main/java/com/codenjoy/dojo/services/Command.java
package com.codenjoy.dojo.services;

/**
 * Имплементит возможные направления движения чего либо
 */
public enum Command {
    LEFT,
    RIGHT,
    DOWN,
    ROTATE_CLOCKWISE_90("ACT"),
    ROTATE_CLOCKWISE_180("ACT(2)"),
    ROTATE_CLOCKWISE_270("ACT(3)"),
    SUICIDE("ACT(0,0)");

    private final String text;

    Command() {
        this.text = this.name();
    }

    Command(String text) {
        this.text = text;
    }

    public String toString() {
        return this.text;
    }

    /**
     * @param dice Given dice.
     * @return Random direction for given dice.
     */
    public static Command random(Dice dice) {
        return Command.values()[dice.next(4)];
    }

    /**
     * Возвращает пустую цепочку команд
     * 
     * @return цепочка команд
     */
    public static CommandChain newChain() {
        return new CommandChain();
    }

    /**
     * Возвращает пустую цепочку команд с одной командой, от которой вызван метод
     * 
     * @return цепочка команд
     */
    public CommandChain asChain() {
        CommandChain chain = new CommandChain();
        return chain.then(this);
    }

    /**
     * Возвращает пустую цепочку команд с двумя командами: первая - от которой
     * вызван метод
     * 
     * @param next вторая команда в цепочке
     * @return цепочка команд
     */
    public CommandChain then(Command next) {
        return this.asChain().then(next);
    }
}